### PR TITLE
A Mul containing a Sum now factors correctly

### PR DESCRIPTION
**Pull request by zanzibar7**
Partially solves issue #7902

### CODE REVIEW

**Comment by smichr** [comment]
What happens with `(5*Sum(x*y,(x,1,3))).factor()`?

**Comment by zanzibar7** [comment]
```
In [1]: (5*Sum(x*y,(x,1,3))).factor()
Out[1]:
      3
     ___
     ╲
      ╲   x
5⋅y⋅  ╱
     ╱
     ‾‾‾
    x = 1
```

File: sympy/polys/polytools.py [hunk 1]
```diff
@@ -5620,6 +5620,14 @@ def _symbolic_factor_list(expr, opt, method):
             if base.is_Number:
                 factors.append((base, exp))
                 continue
+        elif hasattr(arg, '_eval_factor'):
```
Comment (smichr) [inline sympy/polys/polytools.py]: I think for consistency in processing I would just make this:

```python
args = list(Mul.make_args(expr))
for arg in args
    ...
    elif hasattr...
        arg = arg._eval_factor()
        if arg.is_Mul:
            args.extend(arg.args)
            continue
    base, exp = arg, S.One  # XXX notice that the else is removed and this line dedented to catch the arg from above
```

Comment (zanzibar7) [inline sympy/polys/polytools.py]: Dropping the `else' leads to a collision for an earlier part of the `if`.  But the general idea is helpful.

**Comment by zanzibar7** [comment]
@smichr In this last commit, I've reorganized the if statement along with your suggestions.

File: sympy/polys/polytools.py [hunk 2]
```diff
@@ -5611,10 +5611,18 @@ def _symbolic_factor_list(expr, opt, method):
...
     coeff, factors = S.One, []
 
-    for arg in Mul.make_args(expr):
+    args = list(Mul.make_args(expr))
+    for arg in args:
         if arg.is_Number:
             coeff *= arg
             continue
+        elif hasattr(arg, '_eval_factor'):
+            for i in Mul.make_args(arg._eval_factor()):
```
Comment (smichr) [inline sympy/polys/polytools.py]: I'm pretty sure you don't want to do it like this. If you have an unsuccessful factoring, you are putting the result on the stack and will end up with an infinite loop. This is the full rendition of what I intended:


```python
def _symbolic_factor_list(expr, opt, method):
    """Helper function for :func:`_symbolic_factor`. """
    coeff, factors = S.One, []

    args = list(Mul.make_args(expr))
    for arg in args:
        if arg.is_Number:
            coeff *= arg
            continue
        if arg.is_Pow:
            base, exp = arg.args
            if base.is_Number:
                factors.append((base, exp))
                continue
        if hasattr(arg, '_eval_factor'):
            arg = arg._eval_factor()
            if arg.is_Mul:
                args.extend(arg.args)
                continue
        base, exp = arg, S.One

        try:
            ...
```

The reason I do it like that is there is some pretty non-trivial code that happens to each args: by just putting factors on the `args` stack you let them be processed consistently without duplicating efforts (other than refactoring an object with the _eval_factor method unsuccessfully the 2nd pass through).

Comment (smichr) [inline sympy/polys/polytools.py]: Here's an example expression:

```
>>> factor(3*y*Sum(x*y**2*3,(x,1,10)))
9*y**3*Sum(x, (x, 1, 10))
```

Comment (zanzibar7) [inline sympy/polys/polytools.py]: In your version, `base, exp` from the second if statement get overwritten by `arg, S.One`.  Tests fail as a result.  Instead, how about

```python
def _symbolic_factor_list(expr, opt, method):
    """Helper function for :func:`_symbolic_factor`. """
    coeff, factors = S.One, []

    args = list(Mul.make_args(expr))
    for arg in args:
        if arg.is_Number:
            coeff *= arg
            continue
        if hasattr(arg, '_eval_factor'):
            arg = arg._eval_factor()
			if arg.is_Mul:
				args.extend(arg.args)
				continue
        if arg.is_Pow:
            base, exp = arg.args
            if base.is_Number:
                factors.append((base, exp))
                continue
        else:
            base, exp = arg, S.One

        try: ...
```

Comment (smichr) [inline sympy/polys/polytools.py]: > Instead, how about

That looks better and tests agree! :-) +1

**Comment by smichr** [comment]
Sorry! I committed without realizing you didn't make the change internally. Will submit a new PR with the changes and commit if tests pass.

**Pull request merged**